src/rt/sections_osx_x86_64.d

```diff
@@ -114,11 +113,62 @@ void scanTLSRanges(void[] rng, scope void delegate(void* pbeg, void* pend) nothr
 
 private:
 
-extern(C) void _d_dyld_getTLSRange(void*, void**, size_t*);
+// Declarations from dyld_priv.h in dyld, available on 10.7+.
+enum dyld_tlv_states
+{
+    allocated = 10,
+    deallocated = 20
+}
+
+struct dyld_tlv_info
+{
+    size_t info_size;
+    void * tlv_addr;
+    size_t tlv_size;
+}
+
+alias dyld_tlv_state_change_handler = Block!(void, dyld_tlv_states, dyld_tlv_info*)*;
+extern(C) void dyld_enumerate_tlv_storage(dyld_tlv_state_change_handler handler);
 
-__gshared SectionGroup _sections;
 ubyte dummyTlsSymbol;
 
+struct TLSRange
+{
+    void* start;
+    size_t size;
+
+    bool isValid()
+    {
+        return start !is null && size > 0;
+    }
+
+    void[] toArray()
+    {
+        return start[0 .. size];
+    }
+}
+
+TLSRange getTLSRange()
+{
+    void* tlsSymbol = &dummyTlsSymbol;
+    TLSRange range;
+
+    auto handler = block((dyld_tlv_states state, dyld_tlv_info* info){
+        assert(state == dyld_tlv_states.allocated);
+        if (info.tlv_addr <= tlsSymbol && tlsSymbol <
+            (info.tlv_addr + info.tlv_size)
+        )
+        {
+            range = TLSRange(info.tlv_addr, info.tlv_size);
+        }
+    });
+
+    dyld_enumerate_tlv_storage(&handler);
```

Comment:
This handler doesn't escape, does it ?
In which case the delegate literal at line 156 should be made a `scope` delegate to avoid needless memory allocation.
What is blocks semantic in that regard ? Is it mostly / only used for callback or can it be returned as a true delegate as well ?